File: backend/app/schemas/voice_library.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.voice_library import (
    VoiceImportJobStatus,
    VoiceImportTaskStatus,
    VoiceItemFieldStatus,
    VoiceSourceChannel,
)


class VoiceLibraryItemBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    reference_text: str | None = None
    audio_file_path: str | None = None
    is_enabled: bool = True


class VoiceLibraryItemCreate(VoiceLibraryItemBase):
    pass


class VoiceLibraryItemUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=255)
    reference_text: str | None = None
    audio_file_path: str | None = None
    is_enabled: bool | None = None


class VoiceLibraryItemResponse(BaseModel):
    id: int
    name: str
    reference_text: str | None = None
    audio_file_path: str | None = None
    audio_url: str | None = None
    has_audio: bool = False
    is_enabled: bool
    is_builtin: bool
    source_channel: VoiceSourceChannel | None = None
    auto_parse_text: bool = True
    source_url: str | None = None
    source_file_name: str | None = None
    source_post_id: str | None = None
    source_post_updated_at: str | None = None
    clip_start_requested_sec: float | None = None
    clip_end_requested_sec: float | None = None
    clip_start_actual_sec: float | None = None
    clip_end_actual_sec: float | None = None
    name_status: VoiceItemFieldStatus
    reference_text_status: VoiceItemFieldStatus
    processing_stage: str | None = None
    processing_message: str | None = None
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime


class VoiceLibraryListResponse(BaseModel):
    items: list[VoiceLibraryItemResponse]
    total: int
    page: int | None = None
    page_size: int | None = None


class VoiceLibraryImportAudioFilesResponse(BaseModel):
    item_ids: list[int] = Field(default_factory=list)


class VoiceLibraryImportAudioRow(BaseModel):
    index: int = Field(..., ge=0)
    name: str | None = Field(default=None, max_length=255)
    auto_parse_text: bool = True


class VoiceLibraryImportVideoLinkRequest(BaseModel):
    url: str = Field(..., min_length=1)
    start_time: str | None = None
    end_time: str | None = None


class VoiceLibraryImportVideoLinkCreateResponse(BaseModel):
    job_id: str
    job_ids: list[str] = Field(default_factory=list)
    item_ids: list[int] = Field(default_factory=list)


class VoiceLibraryImportTaskResponse(BaseModel):
    id: int
    source_channel: VoiceSourceChannel
    source_url: str | None = None
    source_file_name: str | None = None
    auto_parse_text: bool = True
    clip_start_requested_sec: float | None = None
    clip_end_requested_sec: float | None = None
    clip_start_actual_sec: float | None = None
    clip_end_actual_sec: float | None = None
    status: VoiceImportTaskStatus
    stage: str | None = None
    stage_message: str | None = None
    error_message: str | None = None
    cancel_requested: bool = False
    cancel_requested_at: datetime | None = None
    cancel_reason: str | None = None
    retry_of_task_id: int | None = None
    retry_no: int = 0
    voice_library_item_id: int | None = None
    source_post_id: str | None = None
    source_post_updated_at: str | None = None


class VoiceLibraryImportJobResponse(BaseModel):
    id: str
    status: VoiceImportJobStatus
    total_count: int
    completed_count: int
    success_count: int
    failed_count: int
    canceled_count: int = 0
    cancel_requested: bool = False
    cancel_requested_at: datetime | None = None
    cancel_requested_by: str | None = None
    terminal_at: datetime | None = None
    error_message: str | None = None
    tasks: list[VoiceLibraryImportTaskResponse]
    created_at: datetime
    updated_at: datetime


class VoiceLibraryCancelResponse(BaseModel):
    affected_jobs: int = 0
    affected_tasks: int = 0

File: frontend/src/types/voice-library.ts
export interface VoiceLibraryItem {
  id: number
  name: string
  reference_text?: string | null
  audio_file_path?: string | null
  audio_url?: string | null
  has_audio: boolean
  is_enabled: boolean
  is_builtin: boolean
  source_channel?: 'audio_with_text' | 'audio_file' | 'video_link' | 'builtin' | null
  auto_parse_text: boolean
  source_url?: string | null
  source_file_name?: string | null
  source_post_id?: string | null
  source_post_updated_at?: string | null
  clip_start_requested_sec?: number | null
  clip_end_requested_sec?: number | null
  clip_start_actual_sec?: number | null
  clip_end_actual_sec?: number | null
  name_status: 'pending' | 'running' | 'ready' | 'failed' | 'canceled'
  reference_text_status: 'pending' | 'running' | 'ready' | 'failed' | 'canceled'
  processing_stage?: string | null
  processing_message?: string | null
  error_message?: string | null
  created_at: string
  updated_at: string
}

export interface VoiceLibraryListResponse {
  items: VoiceLibraryItem[]
  total: number
  page?: number | null
  page_size?: number | null
}

export interface VoiceLibraryImportAudioFilesResponse {
  item_ids: number[]
}

export interface VoiceLibraryImportAudioRow {
  index: number
  name?: string | null
  auto_parse_text: boolean
}

export interface VoiceLibraryImportTask {
  id: number
  source_channel: 'audio_with_text' | 'audio_file' | 'video_link' | 'builtin'
  source_url?: string | null
  source_file_name?: string | null
  auto_parse_text: boolean
  clip_start_requested_sec?: number | null
  clip_end_requested_sec?: number | null
  clip_start_actual_sec?: number | null
  clip_end_actual_sec?: number | null
  status: 'pending' | 'running' | 'completed' | 'failed' | 'canceled'
  stage?: string | null
  stage_message?: string | null
  error_message?: string | null
  cancel_requested?: boolean
  cancel_requested_at?: string | null
  cancel_reason?: string | null
  retry_of_task_id?: number | null
  retry_no?: number
  voice_library_item_id?: number | null
  source_post_id?: string | null
  source_post_updated_at?: string | null
}

export interface VoiceLibraryImportJob {
  id: string
  status: 'pending' | 'running' | 'completed' | 'failed' | 'canceled'
  total_count: number
  completed_count: number
  success_count: number
  failed_count: number
  canceled_count?: number
  cancel_requested?: boolean
  cancel_requested_at?: string | null
  cancel_requested_by?: string | null
  terminal_at?: string | null
  error_message?: string | null
  tasks: VoiceLibraryImportTask[]
  created_at: string
  updated_at: string
}

export interface VoiceLibraryImportVideoLinkCreateResponse {
  job_id: string
  job_ids?: string[]
  item_ids: number[]
}

export interface VoiceLibraryCancelResponse {
  affected_jobs: number
  affected_tasks: number
}
